k2/torch/csrc/fsa_class.cu

```diff
@@ -421,70 +427,61 @@ void RaggedArc::SetAttr(const std::string &name, py::object value) {
 
   all_attr_names.insert(name);
 
-  if (THPVariable_Check(value.ptr())) {
-    torch::Tensor tensor = value.cast<torch::Tensor>();
+  if (value.isTensor()) {
+    torch::Tensor tensor = value.toTensor();
     SetAttr(name, tensor);
     return;
   }
 
-  try {
-    RaggedAny ragged_tensor = value.cast<RaggedAny>();
-    SetAttr(name, ragged_tensor);
+  if (value.isCustomClass()) {
+    torch::intrusive_ptr<RaggedAnyHolder> ragged_any_holder =
+        value.toCustomClass<RaggedAnyHolder>();
+    SetAttr(name, *(ragged_any_holder->ragged));
     return;
-  } catch (const py::cast_error &) {
-    // do nothing.
   }
 
   all_attr_names.insert(name);
   other_attrs[name] = value;
 }
 
-py::object RaggedArc::GetAttr(const std::string &name) const {
+torch::IValue FsaClass::GetAttr(const std::string &name) const {
   if (name == "grad") {
-    return py::cast(Scores().grad());
+    return torch::IValue(Scores().grad());
   }
 
   if (!HasAttr(name)) {
     std::ostringstream os;
     os << "No such attribute '" << name << "'";
-    // It's safe to use c_str() here as it is copied inside PyErr_SetString()
-    //
-    // See https://github.com/python/cpython/blob/main/Python/errors.c#L234
-    PyErr_SetString(PyExc_AttributeError, os.str().c_str());
-    throw py::error_already_set();
+    throw std::runtime_error(os.str().c_str());
   }
 
   {
     auto it = tensor_attrs.find(name);
     if (it != tensor_attrs.end()) {
-      return py::cast(it->second);
+      return torch::IValue(it->second);
     }
   }
 
   {
     auto it = ragged_tensor_attrs.find(name);
     if (it != ragged_tensor_attrs.end()) {
-      return py::cast(it->second);
+      return torch::make_custom_class<RaggedAnyHolder>(
+          std::make_shared<RaggedAny>(it->second));
     }
   }
 
   return other_attrs.at(name);
 }
 
-void RaggedArc::DeleteAttr(const std::string &name) {
+void FsaClass::DeleteAttr(const std::string &name) {
   {
     auto it = all_attr_names.find(name);
     if (it != all_attr_names.end()) {
       all_attr_names.erase(it);
     } else {
       std::ostringstream os;
       os << "No such attribute '" << name << "'";
-      // It's safe to use c_str() here as it is copied inside
-      // PyErr_SetString()
-      //
-      // See https://github.com/python/cpython/blob/main/Python/errors.c#L234
-      PyErr_SetString(PyExc_AttributeError, os.str().c_str());
-      throw py::error_already_set();
+      throw std::runtime_error(os.str().c_str());
     }
   }
 
```

Comment:
Please move the block between line 488 and line 494 to line 503.

You don't always need to delete the fillers. Only when the attribute to be deleted is a tensor attribute, do you need to check for its filler attribute.